spanner/api/Spanner/Program.cs

```diff
@@ -1343,6 +1378,177 @@ await connection.CreateDdlCommand($"DROP TABLE {table}")
+
+                for (int i = 0; i < 100; i++)
+                {
+                    using (var tx = await connection.BeginTransactionAsync())
+                    using (var cmd = connection.CreateInsertCommand("Singers", new SpannerParameterCollection
+                    {
+                        {"SingerId", SpannerDbType.String},
+                        {"FirstName", SpannerDbType.String},
+                        {"LastName", SpannerDbType.String}
+                    }))
+                    {
+                        cmd.Transaction = tx;
+                        for (var x = 1; x < 2500; x++)
+                        {
+                            maxSingerId++;
+                            string nameSuffix = Guid.NewGuid().ToString().Substring(0, 8);
+                            cmd.Parameters["SingerId"].Value = maxSingerId;
+                            cmd.Parameters["FirstName"].Value = $"FirstName-{nameSuffix}";
+                            cmd.Parameters["LastName"].Value = $"LastName-{nameSuffix}";
+                            cmd.ExecuteNonQuery();
+                        }
+                        await tx.CommitAsync();
+                    }
+                }
+            }
+            Console.WriteLine("Done inserting sample records...");
+        }
+
+        // [START spanner_batch_client]
+        public static object BatchReadRecords(string projectId,
+             string instanceId, string databaseId)
+        {
+            var response =
+                DistributedReadAsync(projectId, instanceId, databaseId);
+            Console.WriteLine("Waiting for operation to complete...");
+            response.Wait();
+            Console.WriteLine($"Operation status: {response.Status}");
+            return ExitCode.Success;
+        }
+
+        private static async Task DistributedReadAsync(string projectId,
+            string instanceId, string databaseId)
+        {
+            string connectionString =
+                $"Data Source=projects/{projectId}/instances/{instanceId}"
+                + $"/databases/{databaseId}";
+            using (var connection = new SpannerConnection(connectionString))
+            {
+                await connection.OpenAsync();
+
+                using (var transaction =
+                    await connection.BeginReadOnlyTransactionAsync())
+                using (var cmd =
+                    connection.CreateSelectCommand(
+                        "SELECT SingerId, FirstName, LastName FROM Singers"))
+                {
+                    transaction.DisposeBehavior =
+                        DisposeBehavior.CloseResources;
+                    cmd.Transaction = transaction;
+                    var partitions = await cmd.GetReaderPartitionsAsync();
+                    var transactionId = transaction.TransactionId;
+                    await Task.WhenAll(partitions.Select(
+                            x => DistributedReadWorkerAsync(x, transactionId)))
```

Comment:
Here you call a method the number of times as the number of partitions. Each method call will create a separate connection and transaction within outer connection and transaction. Is that really what you want?  

Also, the static variables - do you really need them? Can you just define them in the calling method and pass them by reference to the method that is incrementing their value? This way you don't risk the possibility that someone will use them in a wrong way.

Reply:
Yes we want the partition reads to run in parallel controlled by the calling transaction.

The static variables simplify the method signature for inclusion in documentation. I'd prefer to leave as is.